lib/manageiq/appliance_console/postgres_admin.rb

```diff
@@ -320,6 +331,84 @@ class << self
     ensure
       File.delete(error_path) if File.exist?(error_path)
     end
+
+    private_class_method def self.prepare_restore(backup_type, dbname)
+      if application_connections?
+        message = "Database restore failed. Shut down all evmserverd processes before attempting a database restore"
+        ManageIQ::ApplianceConsole.logger.error(message)
+        raise message
+      end
+
+      disable_replication(dbname)
+
+      conn_count = connection_count(backup_type, dbname)
+      if conn_count > 1
+        message = "Database restore failed. #{conn_count - 1} connections remain to the database."
+        ManageIQ::ApplianceConsole.logger.error(message)
+        raise message
+      end
+    end
+
+    private_class_method def self.application_connections?
+      result = [{"count" => 0}]
+
+      with_pg_connection do |conn|
+        result = conn.exec("SELECT COUNT(pid) FROM pg_stat_activity WHERE application_name LIKE '%MIQ%'")
+      end
+
+      result[0]["count"].to_i > 0
+    end
+
+    private_class_method def self.disable_replication(dbname)
+      require 'pg/logical_replication'
+
+      with_pg_connection do |conn|
+        pglogical = PG::LogicalReplication::Client.new(conn)
+
+        if pglogical.subscriber?
+          pglogical.subcriptions(dbname).each do |subscriber|
+            sub_id = subscriber["subscription_name"]
+            begin
+              pglogical.drop_subscription(sub_id, true)
+            rescue PG::InternalError => e
+              raise unless e.message =~ /could not connect to publisher/
+              raise unless e.message =~ /replication slot .* does not exist/
+
+              pglogical.disable_subscription(sub_id).check
+              pglogical.alter_subscription_options(sub_id, "slot_name" => "NONE")
+              pglogical.drop_subscription(sub_id, true)
+            end
+          end
+        elsif pglogical.publishes?('miq')
+          pglogical.drop_publication('miq')
+        end
+      end
+    end
+
+    private_class_method def self.connection_count(backup_type, dbname)
+      result = nil
+
+      with_pg_connection do |conn|
+        query  = "SELECT COUNT(pid) FROM pg_stat_activity"
+        query << " WHERE backend_type = 'client backend'" if backup_type == :basebackup
+        query << " WHERE datname = '#{dbname}'"           if backup_type == :pgdump
+        result = conn.exec(query)
+      end
+
+      result = conn.result[0]["count"].to_i
```

Comment:
result is set on line 395, but then it's not used here.  Should this be something like?

```suggestion
      result = result[0]["count"].to_i if result
```

Reply:
In fact, conn is not available here either, so as-is this can't work.

Reply:
Yeah, I fixed this prior to seeing this comment.  There was a decent amount wrong here that I am currently in the process of fixing (hence the current test failures).